### PR TITLE
Moved blpapi-wrapper code to blpapi-node project

### LINKED ISSUE #80: Migrate blpapi-wrapper into blpapi-node project.
Migrate `blpapi-wrapper.ts` into the [blpapi-node](http://github.com/bloomberg/blpapi-node) project as a higher-level user-facing API.  Some things to consider:
- Should the existing `blpapi-node` API exist alongside this higher-level API?
- How much of this functionality should be pushed down into the native/binding layer?
- What's the strategy for migrating to depend upon NodeJS 0.12.x features (i.e., generators)?

### PR DESCRIPTION
**Pull request by thmiceli**
Closes #80

Also updated tests.
There is still an issue with the installation of the blpapi module. For now after running 'npm install' in the blpapi-http root dir, you'll have to 'cd' to 'node_modules/blpapi' and run 'npm install' there.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Bloomberg Open API HTTP Proxy",
   "main": "server.js",
   "dependencies": {
-    "blpapi": "^1.2.0",
+    "blpapi": "thmiceli/blpapi-node#move-wrapper-to-node",
```
Comment (ericvw) [inline package.json]: Should this be the `bloomberg/blpapi-node#node-v0.12` branch?

Comment (thmiceli) [inline package.json]: Until the corresponding PR for blpapi-node has been merged we will need this for CI to succeed.

Comment (ericvw) [inline package.json]: I merged the corresponding PR in `blpapi-node`.

File: lib/interface.ts [hunk 2]
```diff
@@ -52,7 +51,8 @@ export interface IOurRequest extends restify.Request {
     clientCert?: any;
     blpSession: BAPI.Session;
     apiSession?: IAPISession;
-    identity?: blpapi.Identity;
+    identity?: BAPI.IIdentity;
+    label?: string;
```
Comment (ericlu88) [inline lib/interface.ts]: I don't think we are exposing the `label` in the http project yet. Let's leave it out until we have a real usage of it.

File: lib/middleware/blp-session.ts [hunk 3]
```diff
@@ -33,22 +33,20 @@ function createSession(sessionId: string): Promise<blpapi.Session>
...
-                return s.authorize();
-            });
+            p = p.then((): Promise<blpapi.IIdentity> => {
+                return s.authenticate().then(s.authorize);
+            }).return(s);
         }
-        // p.return is changing the type of the promise, so we can't store it in the same variable.
-        var p2 = p.return(s);
-        p2.catch((err: Error): any => {  // Use any to make ts happy
+        p.catch((err: Error): any => {  // Use any to make ts happy
```
Comment (ericlu88) [inline lib/middleware/blp-session.ts]: `p = p.catch(...` so we will store the promise that chained with catch block.

File: lib/middleware/request-handler.ts [hunk 4]
```diff
@@ -247,8 +223,17 @@ function onSubscribe(req: Interface.IOurRequest,
...
 
-    // Subscribe user request through blpapi-wrapper
-    req.blpSession.subscribe(subscriptions, req.identity)
+    // Build argument list for call to 'subscribe'.
+    var args: any[] = [subscriptions];
+    // Append optional arguments.
+    if (req.identity) {
+        args.push(req.identity);
+    }
+    if (req.label) {
```
Comment (ericlu88) [inline lib/middleware/request-handler.ts]: This will never happen for now.

File: lib/middleware/request-handler.ts [hunk 5]
```diff
@@ -347,7 +332,13 @@ function onGetToken(req: Interface.IOurRequest,
                     res: Interface.IOurResponse,
                     next: restify.Next): void
 {
-    doRequest('//blp/apiauth', 'AuthorizationTokenRequest', null /* identity */, req, res, next);
+    req.blpSession.authenticate().then((token: string): void => {
+        res.sendChunk(token);
```
Comment (ericlu88) [inline lib/middleware/request-handler.ts]: `res.sendWhole()`?

File: lib/middleware/request-handler.ts [hunk 6]
```diff
@@ -362,8 +353,8 @@ function onAuthorize(req: Interface.IOurRequest,
         return next(new restify.BadRequestError(errorString));
     }
 
-    req.blpSession.authorizeUser(req.body)
-        .then((identity: blpapi.Identity): void => {
+    req.blpSession.authorize(req.body).
```
Comment (ericlu88) [inline lib/middleware/request-handler.ts]: I think `token` should either be a property of `req.body` or as part of url paramaters.

File: lib/middleware/request-handler.ts [hunk 7]
```diff
@@ -492,9 +483,34 @@ export function onRequest(req: Interface.IOurRequest,
...
+    // Build argument list for call to 'request'.
+    var args = [uri,
+                req.query.type,
+                req.body];
+
+    // Append optional arguments (identity, label and callback).
+    if (req.identity) {
+        args.push(req.identity);
+    }
+    if (req.label) {
```
Comment (ericlu88) [inline lib/middleware/request-handler.ts]: same as above.

File: test/lib/mock-blpapi.ts [hunk 8]
```diff
@@ -80,181 +82,202 @@ export class Session extends events.EventEmitter implements ISession {
...
+                    return true;
+                }
+                return false;
+            });
+            if (hasInvalidServiceUri) {
+                reject(new Error('Invalid service name'));
+                return;
+            }
+            subs.forEach((sub: Subscription): void => {
+                var cid = _.findKey(subs, (other: Subscription): boolean => {
```
Comment (ericlu88) [inline test/lib/mock-blpapi.ts]: i don't see an usage of `cid` here.

File: test/lib/mock-blpapi.ts [hunk 9]
```diff
@@ -80,181 +82,202 @@ export class Session extends events.EventEmitter implements ISession {
...
+                    return sub === other;
+                });
+                var serviceUri = getServiceForSecurity(sub.security);
+                var events: string[] = SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri];
+                _.forEach(events, (e: string): void => {
+                    ipc.on(util.format('subscription-%s', e), (): void => {
```
Comment (ericlu88) [inline test/lib/mock-blpapi.ts]: We need a way to differentiate different subscriptions through the ipc channel(`cid`). In the current setup, all subscriptions will emit data if one commands coming through ipc.

File: test/lib/mock-blpapi.ts [hunk 10]
```diff
@@ -264,7 +287,10 @@ export class Session extends events.EventEmitter implements ISession {
                     return true;
```
Comment (ericlu88) [inline test/lib/mock-blpapi.ts]: `this[s].call(this, responseName, callback. cid)`

**Comment by thmiceli** [comment]
I have updated the PR after addressing the review comments.

**Comment by ericlu88** [comment]
LGTM。

**Pull request merged**